homestar-runtime/src/event_handler/notification/network.rs

```diff
@@ -0,0 +1,2671 @@
+            error: error.to_string(),
+        }
+    }
+}
+
+impl DagJson for IncomingConnectionError {}
+
+impl From<IncomingConnectionError> for Ipld {
+    fn from(notification: IncomingConnectionError) -> Self {
+        Ipld::Map(BTreeMap::from([
+            (TIMESTAMP_KEY.into(), notification.timestamp.into()),
+            (ERROR_KEY.into(), notification.error.into()),
+        ]))
+    }
+}
+
+impl TryFrom<Ipld> for IncomingConnectionError {
+    type Error = anyhow::Error;
+
+    fn try_from(ipld: Ipld) -> Result<Self, Self::Error> {
+        let map = from_ipld::<BTreeMap<String, Ipld>>(ipld)?;
+
+        let timestamp = from_ipld(
+            map.get(TIMESTAMP_KEY)
+                .ok_or_else(|| anyhow!("missing {TIMESTAMP_KEY}"))?
+                .to_owned(),
+        )?;
+
+        let error = from_ipld(
+            map.get(ERROR_KEY)
+                .ok_or_else(|| anyhow!("missing {ERROR_KEY}"))?
+                .to_owned(),
+        )?;
+
+        Ok(IncomingConnectionError { timestamp, error })
+    }
+}
+
+#[derive(Debug, Clone, JsonSchema)]
+#[schemars(rename = "discovered_mdns")]
+pub struct DiscoveredMdns {
+    timestamp: i64,
+    #[schemars(description = "Peers discovered by peer ID and multiaddress")]
+    peers: BTreeMap<String, String>,
+}
+
+impl DiscoveredMdns {
+    pub(crate) fn new(peers: BTreeMap<PeerId, Multiaddr>) -> DiscoveredMdns {
+        DiscoveredMdns {
+            timestamp: Utc::now().timestamp_millis(),
+            peers: peers
+                .iter()
+                .map(|(peer_id, address)| (peer_id.to_string(), address.to_string()))
+                .collect(),
+        }
+    }
+}
+
+impl DagJson for DiscoveredMdns {}
+
+impl From<DiscoveredMdns> for Ipld {
+    fn from(notification: DiscoveredMdns) -> Self {
+        let peers: BTreeMap<String, Ipld> = notification
+            .peers
+            .into_iter()
+            .map(|(peer_id, address)| (peer_id, address.into()))
+            .collect();
+
+        let map: BTreeMap<String, Ipld> = BTreeMap::from([
+            (TIMESTAMP_KEY.into(), notification.timestamp.into()),
+            (PEERS_KEY.into(), peers.into()),
+        ]);
+
+        Ipld::Map(map)
+    }
+}
+
+impl TryFrom<Ipld> for DiscoveredMdns {
+    type Error = anyhow::Error;
+
+    fn try_from(ipld: Ipld) -> Result<Self, Self::Error> {
+        let map = from_ipld::<BTreeMap<String, Ipld>>(ipld)?;
+
+        let timestamp = from_ipld(
+            map.get(TIMESTAMP_KEY)
+                .ok_or_else(|| anyhow!("missing {TIMESTAMP_KEY}"))?
+                .to_owned(),
+        )?;
+
+        let peers = from_ipld::<BTreeMap<String, String>>(
+            map.get(PEERS_KEY)
+                .ok_or_else(|| anyhow!("missing {PEERS_KEY}"))?
+                .to_owned(),
+        )?;
+
+        Ok(DiscoveredMdns { timestamp, peers })
+    }
+}
+
+#[derive(Debug, Clone, JsonSchema)]
+#[schemars(rename = "discovered_rendezvous")]
+pub struct DiscoveredRendezvous {
```

Comment:
as discussed, having a rendezvous under this would be good in terms of organization, but maybe we issue this kinda movement @bgins.

Reply:
Refactored network notifications into modules per category, rendezvous, mdns, etc.